File: android-native/app/src/main/java/in/videh/filedtracker/nativeapp/ApiClient.java
package in.videh.filedtracker.nativeapp;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public final class ApiClient {
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");

    private static final OkHttpClient SHARED = AppConfig.okHttpBuilder()
            .connectTimeout(12, TimeUnit.SECONDS)
            .readTimeout(45, TimeUnit.SECONDS)
            .writeTimeout(45, TimeUnit.SECONDS)
            .callTimeout(50, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .build();

    /** Shorter per-attempt timeout — failover to backup base quickly. */
    private static final OkHttpClient OTP = AppConfig.okHttpBuilder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(20, TimeUnit.SECONDS)
            .writeTimeout(20, TimeUnit.SECONDS)
            .callTimeout(22, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .build();

    private static final OkHttpClient FACE = AppConfig.okHttpBuilder()
            .connectTimeout(15, TimeUnit.SECONDS)
            .readTimeout(55, TimeUnit.SECONDS)
            .writeTimeout(55, TimeUnit.SECONDS)
            .callTimeout(65, TimeUnit.SECONDS)
            .build();

    public static class ApiError extends Exception {
        public final int statusCode;

        ApiError(int statusCode, String message) {
            super(message);
            this.statusCode = statusCode;
        }
    }

    private final OkHttpClient http;
    private final String apiBase;
    private final String token;

    public ApiClient(Context ctx) {
        this.apiBase = SessionStore.apiBase(ctx);
        this.token = SessionStore.token(ctx);
        this.http = SHARED;
    }

    private static boolean isTransportFailure(Exception e) {
        if (e instanceof ApiError) return false;
        if (e instanceof java.net.SocketTimeoutException) return true;
        if (e instanceof java.io.InterruptedIOException) return true;
        if (e instanceof IOException) return true;
        return false;
    }

    private static ApiError asNetworkError(Exception e, String fallback) {
        String m = e.getMessage() != null ? e.getMessage().toLowerCase() : "";
        if (e instanceof java.net.SocketTimeoutException
                || m.contains("timeout")
                || m.contains("timed out")) {
            return new ApiError(408, "Network timeout. Check internet and try again.");
        }
        if (m.contains("connection abort")
                || m.contains("connection reset")
                || m.contains("broken pipe")
                || m.contains("unreachable")
                || m.contains("failed to connect")) {
            return new ApiError(0, "Network interrupted. Check internet and try again.");
        }
        return new ApiError(0, fallback);
    }

    /** Try domain then Elastic IP (and any preferred base). */
    private static Response executeFailover(OkHttpClient client, String path, String method, RequestBody body)
            throws IOException, ApiError {
        Exception last = null;
        for (String base : AppConfig.apiBases(AppConfig.API_BASE)) {
            Request.Builder b = new Request.Builder()
                    .url(base + path)
                    .addHeader("X-Client-Source", "native")
                    .addHeader("Content-Type", "application/json")
                    .addHeader("User-Agent", "AAPAttendanceNative/1.4.0");
            if ("GET".equals(method)) b.get();
            else b.method(method, body);
            try {
                return client.newCall(b.build()).execute();
            } catch (Exception e) {
                last = e;
                if (!isTransportFailure(e)) break;
            }
        }
        throw asNetworkError(last != null ? last : new IOException("unreachable"), "Network error. Try again.");
    }

    private Response executeAuthFailover(String path, String method, RequestBody body)
            throws IOException, ApiError {
        Exception last = null;
        for (String base : AppConfig.apiBases(apiBase)) {
            Request.Builder b = new Request.Builder()
                    .url(base + path)
                    .addHeader("Authorization", "Bearer " + token)
                    .addHeader("X-Client-Source", "native")
                    .addHeader("Content-Type", "application/json")
                    .addHeader("User-Agent", "AAPAttendanceNative/1.4.0");
            if ("GET".equals(method)) b.get();
            else b.method(method, body);
            try {
                return http.newCall(b.build()).execute();
            } catch (Exception e) {
                last = e;
                if (!isTransportFailure(e)) break;
            }
        }
        throw asNetworkError(last != null ? last : new IOException("unreachable"), "Network error. Try again.");
    }

    public static JSONObject getAppVersion() throws IOException, ApiError {
        try (Response res = executeFailover(SHARED, "/api/app/version", "GET", null)) {
            String body = res.body() != null ? res.body().string() : "";
            if (!res.isSuccessful()) {
                throw new ApiError(res.code(), "Could not check for updates");
            }
            return new JSONObject(body);
        } catch (ApiError e) {
            throw e;
        } catch (JSONException e) {
            throw new IOException(e);
        }
    }

    private JSONObject readJson(Response res) throws IOException, ApiError, JSONException {
        String body = res.body() != null ? res.body().string() : "";
        if (!res.isSuccessful()) {
            String err = body;
            try {
                JSONObject o = new JSONObject(body);
                err = o.optString("error", body);
            } catch (Exception ignored) {
            }
            if (res.code() == 401) {
                throw new ApiError(401, "Session expired. Please log in again.");
            }
            throw new ApiError(res.code(), err.isEmpty() ? "Request failed (" + res.code() + ")" : err);
        }
        if (body.isEmpty()) return new JSONObject();
        return new JSONObject(body);
    }

    public static void requestOtp(String phone) throws IOException, ApiError {
        JSONObject body = new JSONObject();
        try {
            body.put("phone", phone);
        } catch (Exception e) {
            throw new IOException(e);
        }
        try (Response res =
                executeFailover(OTP, "/api/auth/otp/request", "POST", RequestBody.create(body.toString(), JSON))) {
            if (!res.isSuccessful()) {
                String msg = res.body() != null ? res.body().string() : "";
                try {
                    msg = new JSONObject(msg).optString("error", msg);
                } catch (Exception ignored) {
                }
                throw new ApiError(
                        res.code(), msg.isEmpty() ? "Could not send OTP. Check network and try again." : msg);
            }
        } catch (ApiError e) {
            throw e;
        }
    }

    public static JSONObject verifyOtp(String phone, String otp) throws IOException, ApiError, JSONException {
        JSONObject body = new JSONObject();
        try {
            body.put("phone", phone);
            body.put("otp", otp);
        } catch (Exception e) {
            throw new IOException(e);
        }
        try (Response res =
                executeFailover(OTP, "/api/auth/otp/verify", "POST", RequestBody.create(body.toString(), JSON))) {
            String raw = res.body() != null ? res.body().string() : "";
            if (!res.isSuccessful()) {
                String msg = raw;
                try {
                    msg = new JSONObject(raw).optString("error", raw);
                } catch (Exception ignored) {
                }
                throw new ApiError(res.code(), msg.isEmpty() ? "OTP verification failed" : msg);
            }
            return new JSONObject(raw);
        }
    }

    public JSONObject getMe() throws IOException, ApiError {
        try (Response res = executeAuthFailover("/api/me", "GET", null)) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject getAttendance() throws IOException, ApiError {
        try (Response res = executeAuthFailover("/api/attendance", "GET", null)) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject punchIn(double lat, double lng, Double accuracy, JSONArray descriptor, String image)
            throws IOException, ApiError {
        JSONObject body = punchBody(lat, lng, accuracy, descriptor, image);
        try (Response res =
                executeAuthFailover("/api/attendance", "POST", RequestBody.create(body.toString(), JSON))) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (java.net.SocketTimeoutException e) {
            throw new ApiError(408, "Punch timed out. Stay on this screen and try again.");
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject punchOut(double lat, double lng, Double accuracy, JSONArray descriptor, String image)
            throws IOException, ApiError {
        JSONObject body = punchBody(lat, lng, accuracy, descriptor, image);
        try (Response res = executeAuthFailover(
                "/api/attendance/punch-out", "POST", RequestBody.create(body.toString(), JSON))) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (java.net.SocketTimeoutException e) {
            throw new ApiError(408, "Punch timed out. Stay on this screen and try again.");
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    private static JSONObject punchBody(double lat, double lng, Double accuracy, JSONArray descriptor, String image)
            throws IOException {
        JSONObject body = new JSONObject();
        try {
            body.put("lat", lat);
            body.put("lng", lng);
            if (accuracy != null) body.put("accuracy", accuracy);
            if (descriptor != null && descriptor.length() > 0) body.put("descriptor", descriptor);
            body.put("image", image);
        } catch (Exception e) {
            throw new IOException(e);
        }
        return body;
    }

    public JSONObject getLeave() throws IOException, ApiError {
        try (Response res = executeAuthFailover("/api/leave", "GET", null)) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject createLeave(String fromDate, String toDate, String reason) throws IOException, ApiError {
        JSONObject body = new JSONObject();
        try {
            body.put("fromDate", fromDate);
            body.put("toDate", toDate);
            body.put("reason", reason);
        } catch (Exception e) {
            throw new IOException(e);
        }
        try (Response res =
                executeAuthFailover("/api/leave", "POST", RequestBody.create(body.toString(), JSON))) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject getHistory() throws IOException, ApiError {
        try (Response res = executeAuthFailover("/api/attendance/history", "GET", null)) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject registerFace(JSONArray descriptor, JSONArray samples, String image, boolean coveredUpper)
            throws IOException, ApiError {
        JSONObject body = new JSONObject();
        try {
            body.put("descriptor", descriptor);
            body.put("samples", samples);
            body.put("image", image);
            body.put("usesTurban", coveredUpper);
        } catch (Exception e) {
            throw new IOException(e);
        }
        try (Response res = executeAuthFailover(
                "/api/face/register", "POST", RequestBody.create(body.toString(), JSON))) {
            return readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }

    public JSONObject describeFace(String imageDataUrl) throws IOException, ApiError {
        JSONObject body = new JSONObject();
        try {
            body.put("image", imageDataUrl);
            body.put("relaxed", true);
            body.put("fast", true);
        } catch (Exception e) {
            throw new IOException(e);
        }
        Exception last = null;
        for (String base : AppConfig.apiBases(apiBase)) {
            Request req = new Request.Builder()
                    .url(base + "/api/face/describe")
                    .addHeader("Authorization", "Bearer " + token)
                    .addHeader("X-Client-Source", "native")
                    .addHeader("Content-Type", "application/json")
                    .addHeader("User-Agent", "AAPAttendanceNative/1.4.0")
                    .post(RequestBody.create(body.toString(), JSON))
                    .build();
            try (Response res = FACE.newCall(req).execute()) {
                return readJson(res);
            } catch (ApiError e) {
                throw e;
            } catch (Exception e) {
                last = e;
                if (!isTransportFailure(e)) break;
            }
        }
        if (last instanceof java.net.SocketTimeoutException) {
            throw new ApiError(408, "Face check timed out. Try again on a better network.");
        }
        throw asNetworkError(last != null ? last : new IOException("unreachable"), "Face check failed. Try again.");
    }

    public void reportLocationPermission(boolean foreground, boolean background) throws IOException, ApiError {
        JSONObject body = new JSONObject();
        try {
            body.put("foreground", foreground);
            body.put("background", background);
            body.put("platform", "android");
        } catch (Exception e) {
            throw new IOException(e);
        }
        try (Response res = executeAuthFailover(
                "/api/me/location-permission", "POST", RequestBody.create(body.toString(), JSON))) {
            readJson(res);
        } catch (ApiError e) {
            throw e;
        } catch (Exception e) {
            throw new IOException(e);
        }
    }
}

File: ios-native/AAPAttendance/ApiClient.swift
import Foundation

struct ApiError: LocalizedError {
    let statusCode: Int
    let message: String
    var errorDescription: String? { message }
}

enum ApiClient {
    static func requestOtp(phone: String) async throws {
        _ = try await postPublic(path: "/api/auth/otp/request", body: ["phone": phone])
    }

    static func verifyOtp(phone: String, otp: String) async throws -> [String: Any] {
        try await postPublic(path: "/api/auth/otp/verify", body: ["phone": phone, "otp": otp])
    }

    static func getMe() async throws -> [String: Any] {
        try await authed(path: "/api/me", method: "GET")
    }

    static func getAttendance() async throws -> [String: Any] {
        try await authed(path: "/api/attendance", method: "GET")
    }

    static func getHistory() async throws -> [String: Any] {
        try await authed(path: "/api/attendance/history", method: "GET")
    }

    static func getLeave() async throws -> [String: Any] {
        try await authed(path: "/api/leave", method: "GET")
    }

    static func createLeave(fromDate: String, toDate: String, reason: String) async throws -> [String: Any] {
        try await authed(path: "/api/leave", method: "POST", body: [
            "fromDate": fromDate,
            "toDate": toDate,
            "reason": reason,
        ])
    }

    static func punchIn(lat: Double, lng: Double, accuracy: Double, descriptor: [Double], image: String) async throws -> [String: Any] {
        try await authed(path: "/api/attendance", method: "POST", body: [
            "lat": lat, "lng": lng, "accuracy": accuracy,
            "descriptor": descriptor, "image": image,
        ])
    }

    static func punchOut(lat: Double, lng: Double, accuracy: Double, descriptor: [Double], image: String) async throws -> [String: Any] {
        try await authed(path: "/api/attendance/punch-out", method: "POST", body: [
            "lat": lat, "lng": lng, "accuracy": accuracy,
            "descriptor": descriptor, "image": image,
        ])
    }

    static func registerFace(descriptor: [Double], samples: [[Double]], image: String) async throws -> [String: Any] {
        try await authed(path: "/api/face/register", method: "POST", body: [
            "descriptor": descriptor,
            "samples": samples,
            "image": image,
            // Soft match for eyes/nose/chin when upper head is covered (no UI label).
            "usesTurban": true,
        ])
    }

    static func describeFace(imageDataUrl: String, fast: Bool = true) async throws -> [String: Any] {
        try await authed(path: "/api/face/describe", method: "POST", body: [
            "image": imageDataUrl,
            "relaxed": true,
            "fast": fast,
        ])
    }

    static func reportLocationPermission(foreground: Bool, background: Bool) async throws {
        _ = try await authed(path: "/api/me/location-permission", method: "POST", body: [
            "foreground": foreground,
            "background": background,
            "platform": "ios",
        ])
    }

    private static func postPublic(path: String, body: [String: Any]) async throws -> [String: Any] {
        try await send(url: AppConfig.apiBase + path, method: "POST", token: nil, body: body)
    }

    private static func authed(path: String, method: String, body: [String: Any]? = nil) async throws -> [String: Any] {
        try await send(url: SessionStore.apiBase + path, method: method, token: SessionStore.token, body: body)
    }

    private static func send(url: String, method: String, token: String?, body: [String: Any]?) async throws -> [String: Any] {
        guard let u = URL(string: url) else { throw ApiError(statusCode: 0, message: "Bad URL") }
        var req = URLRequest(url: u)
        req.httpMethod = method
        req.timeoutInterval = 55
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("native", forHTTPHeaderField: "X-Client-Source")
        req.setValue("AAPAttendance/1.3.0 AAPNative/1", forHTTPHeaderField: "User-Agent")
        if let token, !token.isEmpty {
            req.setValue("Bearer \(token)", forHTTPHeaderField: "Authorization")
        }
        if let body {
            req.httpBody = try JSONSerialization.data(withJSONObject: body)
        }
        let data: Data
        let res: URLResponse
        do {
            (data, res) = try await URLSession.shared.data(for: req)
        } catch {
            let msg = error.localizedDescription.lowercased()
            if msg.contains("timed out") || msg.contains("timeout") {
                throw ApiError(statusCode: 408, message: "Network timeout. Check internet and try again.")
            }
            if msg.contains("connection") || msg.contains("offline") || msg.contains("internet") {
                throw ApiError(statusCode: 0, message: "Network interrupted. Check internet and try again.")
            }
            throw ApiError(statusCode: 0, message: "Network error. Check internet and try again.")
        }
        let code = (res as? HTTPURLResponse)?.statusCode ?? 0
        let raw = String(data: data, encoding: .utf8) ?? ""
        let json = (try? JSONSerialization.jsonObject(with: data) as? [String: Any]) ?? [:]
        if !(200..<300).contains(code) {
            let msg = (json["error"] as? String) ?? (raw.isEmpty ? "Request failed (\(code)). Try again." : raw)
            throw ApiError(statusCode: code, message: msg)
        }
        return json
    }
}

extension Dictionary where Key == String, Value == Any {
    func string(_ key: String) -> String? {
        if self[key] is NSNull { return nil }
        let v = self[key] as? String
        return (v?.isEmpty == false) ? v : nil
    }

    func obj(_ key: String) -> [String: Any]? {
        self[key] as? [String: Any]
    }

    func arr(_ key: String) -> [[String: Any]] {
        (self[key] as? [Any])?.compactMap { $0 as? [String: Any] } ?? []
    }

    func doubles(_ key: String) -> [Double] {
        (self[key] as? [Any])?.compactMap { ($0 as? NSNumber)?.doubleValue } ?? []
    }

    func nestedDoubles(_ key: String) -> [[Double]] {
        (self[key] as? [Any])?.map { row -> [Double] in
            (row as? [Any])?.compactMap { ($0 as? NSNumber)?.doubleValue } ?? []
        } ?? []
    }
}
